src/main/java/eu/europeana/metis/sandbox/service/dataset/AsyncDatasetPublishServiceImpl.java

```diff
@@ -3,75 +3,162 @@
+  private final Executor asyncServiceTaskExecutor;
+  private final OaiHarvester oaiHarvester;
+  private final HarvestService harvestService;
+  private final DatasetService datasetService;
+
+
+  public AsyncDatasetPublishServiceImpl(AmqpTemplate amqpTemplate,
+                                        String createdQueue, String transformationToEdmExternalQueue,
+                                        Executor asyncServiceTaskExecutor, OaiHarvester oaiHarvester,
+                                        HarvestService harvestService, DatasetService datasetService,
+                                        @Value("${sandbox.dataset.max-size}") int maxRecords) {
+    this.amqpTemplate = amqpTemplate;
+    this.createdQueue = createdQueue;
+    this.transformationToEdmExternalQueue = transformationToEdmExternalQueue;
+    this.asyncServiceTaskExecutor = asyncServiceTaskExecutor;
+    this.oaiHarvester = oaiHarvester;
+    this.harvestService = harvestService;
+    this.datasetService = datasetService;
+    this.maxRecords = maxRecords;
+  }
+
+
+  @Override
+  public CompletableFuture<Void> runHarvestOaiAsync(String datasetName, String datasetId,
+                                 Country country, Language language, OaiHarvestData oaiHarvestData) {
+        return CompletableFuture.runAsync(
+                () -> harvestOaiPmh(datasetName, datasetId, country, language, oaiHarvestData), asyncServiceTaskExecutor);
+
+  }
+
+  //If we moved this method to HarvestService while keeping runHarvestOaiAsync method here, it will create a cyclic dependency between
+  //this class and HarvestService. Either we keep as it is or we should figure out a solution
+  private void harvestOaiPmh(String datasetName, String datasetId,
+                            Country country, Language language, OaiHarvestData oaiHarvestData){
+    try (OaiRecordHeaderIterator recordHeaderIterator = oaiHarvester
+            .harvestRecordHeaders(
+                    new OaiHarvest(oaiHarvestData.getUrl(), oaiHarvestData.getMetadataformat(), oaiHarvestData.getSetspec()))) {
+
+      AtomicInteger currentNumberOfIterations = new AtomicInteger();
+
+      Record recordDataEncapsulated = Record.builder()
+              .country(country)
+              .language(language)
+              .datasetName(datasetName)
+              .datasetId(datasetId)
+              .content(new byte[0])
+              .build();
+
+      recordHeaderIterator.forEach(recordHeader -> {
+        OaiHarvestData completeOaiHarvestData = new OaiHarvestData(oaiHarvestData.getUrl(),
+                oaiHarvestData.getSetspec(),
+                oaiHarvestData.getMetadataformat(),
+                recordHeader.getOaiIdentifier());
+        currentNumberOfIterations.getAndIncrement();
+
+        if (currentNumberOfIterations.get() > maxRecords) {
+          datasetService.updateRecordsLimitExceededToTrue(datasetId);
+          currentNumberOfIterations.set(maxRecords);
```

Comment:
And if the counting is after, I think reseting the value could be avoided if the check it >=